File: reference-implementations/hmi2018-demonstrator-snapshot/java/org.eclipse.basyx.onem2m/src/main/java/org/eclipse/basyx/onem2m/clients/OneM2MHttpClient.java
package org.eclipse.basyx.onem2m.clients;

import java.lang.reflect.InvocationTargetException;
import java.math.BigInteger;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import org.apache.commons.lang3.NotImplementedException;
import org.eclipse.basyx.onem2m.clients.OneM2MHttpClientConfig.Protocol;
import org.eclipse.basyx.onem2m.manager.OneM2MResourceManager;
import org.eclipse.basyx.onem2m.util.OneM2MResourceUtil;
import org.eclipse.basyx.onem2m.xml.protocols.Cb;
import org.eclipse.basyx.onem2m.xml.protocols.PrimitiveContent;
import org.eclipse.basyx.onem2m.xml.protocols.Resource;
import org.eclipse.basyx.onem2m.xml.protocols.Rqp;
import org.eclipse.basyx.onem2m.xml.protocols.Rsp;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.api.ContentResponse;
import org.eclipse.jetty.client.api.Request;
import org.eclipse.jetty.client.util.FutureResponseListener;
import org.eclipse.jetty.client.util.StringContentProvider;
import org.eclipse.jetty.http.HttpMethod;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

public class OneM2MHttpClient implements IOneM2MClient {

	public static final int CSE_RETRY_MS = 5000;
	
	protected HttpClient httpClient = new HttpClient();
	protected OneM2MHttpClientConfig config;
	protected int rqiCounter = 0; 
	protected Cb cseBase;
	
	public OneM2MHttpClient(OneM2MHttpClientConfig _config) {
		this.config = _config;				
	}
	
	@Override
	public void start() throws Exception {
		this.start(true);
	}

	/**
	 * Starts the client and actively waits (blocked call) until the CSE is available.
	 */
	@Override
	public void start(boolean waitForCSE) throws Exception {
		this.httpClient.start();
		if (waitForCSE) {
			this.waitForCSE();
		}
	}

	@Override
	public void stop() throws Exception {
		this.httpClient.stop();
	}

	@Override
	public Rqp createDefaultRequest(String path, boolean hierarchical) {
		Rqp rqp = new Rqp();
		String to = (this.config.protocol.equals(Protocol.HTTP) ? "http://" : "https://")
						+ this.config.host + ":" + this.config.port
						+ (hierarchical ? "/"+ this.cseBase.getRn() + "/" : "/")  
						+ path;
		
		rqp.setTo(to);
		rqp.setFr(this.config.from);
		rqp.setRqi(Integer.toString(this.rqiCounter++));
				
		return rqp;
	}

	@Override
	public Rsp send(Rqp rqp) throws InterruptedException, TimeoutException, ExecutionException, ClassNotFoundException, NoSuchMethodException, SecurityException, InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
		
		String mime = "application/json";
		String url = rqp.getTo();
		
		Request request = httpClient.newRequest(url);
		request.header("Accept", mime);
		
		if(rqp.getOp().equals(BigInteger.valueOf(1)) && rqp.getTy() == null) {
			new IllegalArgumentException("Ty must be set if op = 1");
		}
		request.header("Content-Type", rqp.getOp().equals(BigInteger.valueOf(1)) ? (mime + ";ty=" + rqp.getTy().toString()) : mime);		
		request.header("X-M2M-Origin", rqp.getFr());
		request.header("X-M2M-Ri", rqp.getRqi());
				
	    if (rqp.getDrt() != null) {
	        request.param("drt", rqp.getDrt().toString());
	    }
	    if (rqp.getFc() != null) {
	        // filter criteria
	        if (rqp.getFc().getFu() != null) {
	        	request.param("fu", rqp.getFc().getFu().toString());
	        }
	        if (rqp.getFc().getLvl() != null) {
	        	request.param("lvl", rqp.getFc().getLvl().toString());
	        }
	        for (String lbl : rqp.getFc().getLbl()) {
	        	request.param("lbl", lbl);
	        }
	    }
		
	    if (rqp.getRcn() != null) {
	        request.param("rcn", rqp.getRcn().toString());
	    }
	    		
		switch (rqp.getOp().intValue()) {
		case 1: { // Create
			request.method(HttpMethod.POST);
			request.content(new StringContentProvider(this.createPayload(rqp.getPc())));
			break;
			}
			
		case 2: { // Retrieve
			request.method(HttpMethod.GET);			
			break;
			}

		case 3: { // Update
			throw new NotImplementedException("Update not implemented yet");
			//break;
			}

		case 4: { // Delete
			request.method(HttpMethod.DELETE);						
			break;
			}

		case 5: { // Notify
			// not implemented
			throw new NotImplementedException("Notify not implemented yet");
			//break;
			}
		
		default:
			throw new IllegalArgumentException("Unknown operation type");
		}
		
		
		
		FutureResponseListener listener = new FutureResponseListener(request, 32 * 1024 * 1024); // 32MB Limit
		request.send(listener);
		ContentResponse response = listener.get(5, TimeUnit.SECONDS);
		//ContentResponse response = request.send(); old code (default limit 2MB)		
		Rsp rsp = this.createRsp(response);		
		return rsp;
	}
	
	protected Rsp createRsp(ContentResponse response) throws ClassNotFoundException, NoSuchMethodException, SecurityException, InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
		Rsp rsp = new Rsp();
		String content = response.getContentAsString();
		if (content.length() > 0 && 
				(response.getStatus() >= 200 &&  response.getStatus() < 300
				)) { 
			JsonObject json = new JsonParser().parse(content).getAsJsonObject();
			Set<Entry<String, JsonElement>> entrySet = json.entrySet();
	        for(Map.Entry<String,JsonElement> entry : entrySet){

	        	if (OneM2MResourceUtil.isResource(entry.getKey())) {
		        	JsonObject jsonObj = entry.getValue().getAsJsonObject();
		        	
		        	String className = entry.getKey().substring("m2m:".length());
		        	String fullClassName = "org.eclipse.basyx.onem2m.xml.protocols." + className.substring(0, 1).toUpperCase() + className.substring(1);
		        	Class<?> c = Class.forName(fullClassName);
		        	
		        	Object o = new Gson().fromJson(jsonObj, c);
		        	
		        	for (Entry<String, JsonElement> childEntry : jsonObj.entrySet()) {
		        		if (childEntry.getKey().startsWith("m2m:")) {
		        			// child found
				        	String childClassName = childEntry.getKey().substring("m2m:".length());
				        	String childFullClassName = "org.eclipse.basyx.onem2m.xml.protocols." + childClassName.substring(0, 1).toUpperCase() + childClassName.substring(1);
				        	Class<?> childClass = Class.forName(childFullClassName);
				        	if (childEntry.getValue() instanceof JsonObject) {
				        		Object child = new Gson().fromJson(childEntry.getValue(), childClass);
				        		OneM2MResourceUtil.addChildToParent((Resource) o, (Resource) child);
				        	} else if (childEntry.getValue() instanceof JsonArray) {
				        		JsonArray arr = (JsonArray) childEntry.getValue();
				        		for (JsonElement element : arr) {
					        		Object child = new Gson().fromJson(element, childClass);
					        		OneM2MResourceUtil.addChildToParent((Resource) o, (Resource) child);				        			
				        		}
				        	}
		        		}
		        	}
		        	
		        	rsp.setPc(new PrimitiveContent());
		        	rsp.getPc().getAnyOrAny().add(o);
	        	} else if (OneM2MResourceUtil.isOtherPrimitiveContent(entry.getKey())) {
		        	Object o = OneM2MResourceUtil.createOtherPrimitiveContent(entry.getKey(), entry.getValue());
		        	rsp.setPc(new PrimitiveContent());
		        	rsp.getPc().getAnyOrAny().add(o);
	        	}
	        }
		}
        rsp.setFr(response.getHeaders().get("x-m2m-origin"));
        rsp.setRqi(response.getHeaders().get("x-m2m-ri"));
        rsp.setRsc(new BigInteger(response.getHeaders().get("x-m2m-rsc")));
		return rsp;
	}
	
	private String createPayload(PrimitiveContent pc) {
		if (pc.getAnyOrAny().size() == 1) {
			Resource resource = (Resource) pc.getAnyOrAny().iterator().next();
	        String payload = "{\"" + OneM2MResourceUtil.getXsdTypeNameFromResource(resource) + "\":"+ new Gson().toJson(resource) + "}";
	        return payload;			
		}
		throw new NotImplementedException("Create payload from array not implemented yet.");
	}
	
	private void waitForCSE(){
		while(this.cseBase == null){
			try{
				this.cseBase = (new OneM2MResourceManager(this).retrieveCSEBase()).getResource();
			}catch (Exception e){
				System.out.println("Cannot connect to CSE, retry in " + (OneM2MHttpClient.CSE_RETRY_MS/1000) + "s");
				try {
					Thread.sleep(OneM2MHttpClient.CSE_RETRY_MS);
				} catch (InterruptedException ie) {
					e.printStackTrace();
				}
			}
		}
	}

	@Override
	public OneM2MHttpClientConfig getConfig() {
		return this.config;
	}
	
}

File: reference-implementations/hmi2018-demonstrator-snapshot/webapp/modules/angular-onem2m/src/clients/onem2m-http-client.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/operator/map'
import { ReplaySubject } from 'rxjs/ReplaySubject';

import { OneM2MClient } from "./onem2m-client";
import { OneM2MResourcesUtil } from '../resources/onem2m-resources-util';

import { Cb, Rsp, Resource, Rqp, PrimitiveContent } from "../resources/onem2m-resources";
import { OneM2MResourceManager } from '../utils/onem2m-resource-manager';

export enum OneM2MHttpClientProtocol {
  HTTP,
  HTTPS
}

export class OneM2MHttpClientConfiguration {
  protocol: OneM2MHttpClientProtocol;
  host: string;
  port: Number;
  from: string;

  constructor(protocol: OneM2MHttpClientProtocol, host: string, port: Number, from: string) {
    this.protocol = protocol;
    this.host = host;
    this.port = port;
    this.from = from;
  }
}

export class OneM2MHttpClient implements OneM2MClient {

  private config: OneM2MHttpClientConfiguration;
  private cb: Cb;
  private rqiCounter = 0;

  constructor(private http: HttpClient) {

  }

  private enum2val(protocol: OneM2MHttpClientProtocol): string {
    if (protocol == OneM2MHttpClientProtocol.HTTP) {
      return "http://";
    } else if (protocol == OneM2MHttpClientProtocol.HTTPS) {
      return "https://";
    }
    return null;
  }

  public getConfiguration(): OneM2MHttpClientConfiguration {
    return this.config;
  }

  public start(config: OneM2MHttpClientConfiguration): Observable<Boolean> {
    let subject = new ReplaySubject<Boolean>(1);
    this.config = config;
    new OneM2MResourceManager(this).retrieveCSEBase().subscribe(
      result => {
        if (result.response.rsc === 2000) {
          this.cb = result.resource;
          subject.next(true);
          subject.complete();
        } else {
          subject.error("Could not retrieve CSE base.");
        }
      },
      error => {
        subject.error(error);
      }
    )
    return subject.asObservable();
  }

  public stop(): void {
    this.config = undefined;
    this.cb = undefined;
  }

  public createDefaultRequest(path: string, hierarchical?: boolean): Rqp {
    var rqp: Rqp = new Rqp();
    rqp.to = this.enum2val(this.config.protocol) + this.config.host + ':' + this.config.port + '/' + ((hierarchical) ? this.cb.rn + '/' : '') + '' + path;
    rqp.fr = this.config.from;
    rqp.rqi = (this.rqiCounter++).toString();

    return rqp;
  }


  public send(rqp: Rqp): Observable<Rsp> {
    if (this.config == null) {
      console.log("Config is not set! Did you call 'start'?");
    }
    var mime: string = "application/json"; // Supports only JSON for now

    // url
    var url = rqp.to;

    // fill header
    var headers: any = {};
    headers['Accept'] = mime;
    if (rqp.op) {
      if (rqp.op === 1 && rqp.ty === undefined) {
        throw new Error("Ty must be set if op = 1");
      }
      headers['Content-Type'] = rqp.op == 1 ? (mime + ';ty=' + rqp.ty) : mime;
    }
    if (rqp.fr) {
      headers['X-M2M-Origin'] = rqp.fr;
    }
    if (rqp.rqi) {
      headers['X-M2M-RI'] = rqp.rqi;
    }

    // fill GET parameters
    var params: string[] = [];

    if (rqp.da) {
      params.push("drt=" + rqp.drt);
    }
    if (rqp.ty) {
      params.push("ty=" + rqp.ty);
    }
    if (rqp.drt) {
      params.push("drt=" + rqp.drt);
    }
    if (rqp.fc !== undefined && rqp.fc !== null) {
      // filter criteria
      if (rqp.fc.fu !== undefined) {
        params.push("fu=" + rqp.fc.fu);
      }
      if (rqp.fc.lbl.length > 0) {
        for (var lbl of rqp.fc.lbl) {
          params.push("lbl=" + lbl);
        }
      }
      if (rqp.fc.lvl != undefined) {
        params.push("lvl=" + rqp.fc.lvl);
      }
    }
    if (rqp.rcn) {
      params.push("rcn=" + rqp.rcn);
    }

    url = params.length > 0 ? url + '?' + params.join('&') : url;

    var obs: Observable<any> = null;
    switch (rqp.op) {
      case 1: { // Create
        if (rqp.pc.anyOrAny.length != 1) {
          throw new Error("Primitive content must have one element!");
        }
        obs = this.http.post(url, rqp.pc.anyOrAny[0], { 'headers': headers, observe: 'response' });
        break;
      }
      case 2: { // Retrieve
        obs = this.http.get(url, { 'headers': headers, observe: 'response' });
        break;
      }
      case 3: { // Update
        break; // not implemented yet        
      }
      case 4: { // Delete
        obs = this.http.delete(url, { 'headers': headers, observe: 'response' });
        break;
      }
      case 5: { // Notify
        break; // not implemented yet        
      }
    }
    var subject: ReplaySubject<Rsp> = new ReplaySubject<Rsp>(1);
    obs.subscribe(
      response => {
        subject.next(this.createRsp(response));
        subject.complete();
      },
      error => {
        subject.error(this.createRsp(error));
      }
    );
    return subject.asObservable();
  }

  private createRsp(response: HttpResponse<any>): Rsp {
    let rsp: Rsp = new Rsp();
    rsp.fr = response.headers.get("x-m2m-origin");
    rsp.rqi = response.headers.get("x-m2m-ri");
    rsp.rsc = Number(response.headers.get("x-m2m-rsc"));
    if (response.statusText.length > 0
      && response.status !== 404
      && response.status !== 409) {
      if (response.body !== null && response.body !== undefined) { // received body
        var json = response.body; // .json();
        if (Object.keys(json).length != 1) {
          throw new Error("Invalid state: A single resource is expected.");
        }
        var type: string = Object.keys(json)[0];
        if (!type.startsWith("m2m:")) {
          throw new Error("Invalid state: Unknown oneM2M response");
        }
        rsp.pc = new PrimitiveContent;
        if (OneM2MResourcesUtil.isResource(type)) {
          var resource: Resource = this.mapResponseJsonToResource(type, json[type]);
          rsp.pc.anyOrAny.push(resource);
        } else if (OneM2MResourcesUtil.isOtherPrimitiveContent(type)) {
          var content: any = this.mapResponseJsonToOtherPrimitveContent(type, json[type]);
          rsp.pc.anyOrAny.push(content);
        }
      }
    }
    return rsp;
  }

  private mapResponseJsonToOtherPrimitveContent(type: string, json: any): Resource {
    var content: any = OneM2MResourcesUtil.createOtherPrimitiveContent(type, json);
    return content;
  }


  private mapResponseJsonToResource(type: string, json: any): Resource {
    var resource: Resource = OneM2MResourcesUtil.createResource(type);
    // fill attributes
    this.mapResponseJsonToResourceProperties(resource, type, json);
    // fill children
    this.mapResponseJsonToResourceChildren(resource, type, json);

    return resource;
  }


  private mapResponseJsonToResourceProperties(resource: Resource, type: string, json: any): void {
    for (var attr of Object.getOwnPropertyNames(json)) {
      if (resource.hasOwnProperty(attr)) {
        resource[attr] = json[attr];
      }
    }
  }
  private mapResponseJsonToResourceChildren(resource: Resource, type: string, json: any): void {
    for (var attr of Object.getOwnPropertyNames(json)) {
      if (attr.startsWith("m2m:")) { // children detected
        var childType: string = attr;
        var childJsons: any[] = json[attr];
        for (var childJson of childJsons) {
          var childResource: Resource = OneM2MResourcesUtil.createResource(childType);
          this.mapResponseJsonToResourceProperties(childResource, childType, childJson);
          if (resource.ri === childResource.pi) {
            this.mapResponseJsonToResourceChildren(childResource, childType, childJson);
            OneM2MResourcesUtil.addChildToParent(resource, childResource);
          }
        }
      }
    }
  }
}
